pkg/typesystem/typesystem_create_test.go

```diff
@@ -0,0 +1,323 @@
+package typesystem
+
+import (
+	"testing"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"github.com/stretchr/testify/require"
+
+	"github.com/openfga/openfga/internal/condition"
+	"github.com/openfga/openfga/pkg/testutils"
+)
+
+func usersetsEquals(t *testing.T, a, b *openfgav1.Usersets) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	childA := a.GetChild()
+	childB := b.GetChild()
+	require.Equal(t, len(childA), len(childB))
+	for i := 0; i < len(childA); i++ {
+		rewriteEquals(t, childA[i], childB[i])
+	}
+}
+
+func tupleToUsersetEquals(t *testing.T, a, b *openfgav1.TupleToUserset) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	cuA := a.GetComputedUserset()
+	cuB := b.GetComputedUserset()
+
+	if cuA == nil || cuB == nil {
+		require.Equal(t, cuA, cuB)
+		return
+	}
+	objectRelationEquals(t, cuA, cuB)
+}
+
+func objectRelationEquals(t *testing.T, a, b *openfgav1.ObjectRelation) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	require.Equal(t, a.GetRelation(), b.GetRelation())
+	require.Equal(t, a.GetObject(), b.GetObject())
+}
+
+func rewriteEquals(t *testing.T, a, b *openfgav1.Userset) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+
+	usersetA := a.GetUserset()
+	usersetB := b.GetUserset()
+
+	switch x := usersetA.(type) {
+	case *openfgav1.Userset_This:
+		_, ok := usersetB.(*openfgav1.Userset_This)
+		require.True(t, ok)
+	case *openfgav1.Userset_ComputedUserset:
+		y, ok := usersetB.(*openfgav1.Userset_ComputedUserset)
+		require.True(t, ok)
+		objectRelationEquals(t, x.ComputedUserset, y.ComputedUserset)
+	case *openfgav1.Userset_TupleToUserset:
+		y, ok := usersetB.(*openfgav1.Userset_TupleToUserset)
+		require.True(t, ok)
+		tupleToUsersetEquals(t, x.TupleToUserset, y.TupleToUserset)
+	case *openfgav1.Userset_Union:
+		y, ok := usersetB.(*openfgav1.Userset_Union)
+		require.True(t, ok)
+		usersetsEquals(t, x.Union, y.Union)
+	case *openfgav1.Userset_Intersection:
+		y, ok := usersetB.(*openfgav1.Userset_Intersection)
+		require.True(t, ok)
+		usersetsEquals(t, x.Intersection, y.Intersection)
+	case *openfgav1.Userset_Difference:
+		y, ok := usersetB.(*openfgav1.Userset_Difference)
+		require.True(t, ok)
+		if x.Difference == nil || y.Difference == nil {
+			require.Equal(t, x.Difference, y.Difference)
+			return
+		}
+		rewriteEquals(t, x.Difference.GetBase(), y.Difference.GetBase())
+		rewriteEquals(t, x.Difference.GetSubtract(), y.Difference.GetSubtract())
+	default:
+		require.Equal(t, usersetA, usersetB)
+	}
+}
+
+func relationReferenceEquals(t *testing.T, a, b *openfgav1.RelationReference) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	require.Equal(t, a.GetCondition(), b.GetCondition())
```

Comment:
Condition is a pointer object.  I suspect you cannot simply assert they are equal.  You may need a separate function to assert name, expression, parameters, metadata etc.

Reply:
`GetCondition` returns a `string`.